Lib/test/datetimetester.py

```diff
@@ -2501,6 +2501,47 @@ def test_strptime(self):
                 self.assertEqual(expected, got)
 
         strptime = self.theclass.strptime
+
+        # bpo-34903: Check that single digit dates and times are allowed.
+        with self.assertRaises(ValueError):
+            # %y does require two digits.
+            newdate = strptime('01/02/3 04:05:06', '%d/%m/%y %H:%M:%S')
+        inputs = [
+            ('%d',
+             '1/02/03 4:5:6', '%d/%m/%y %H:%M:%S', '2003-02-01T04:05:06'),
+            ('%m',
+             '01/2/03 4:5:6', '%d/%m/%y %H:%M:%S', '2003-02-01T04:05:06'),
+            ('%H',
+             '01/02/03 4:05:06', '%d/%m/%y %H:%M:%S', '2003-02-01T04:05:06'),
+            ('%M',
+             '01/02/03 04:5:06', '%d/%m/%y %H:%M:%S', '2003-02-01T04:05:06'),
+            ('%S',
+             '01/02/03 04:05:6', '%d/%m/%y %H:%M:%S', '2003-02-01T04:05:06'),
+            ('%j',
+             '2/03 04am:05:06', '%j/%y %I%p:%M:%S','2003-01-02T04:05:06'),
+            ('%I',
+             '02/03 4am:05:06', '%j/%y %I%p:%M:%S','2003-01-02T04:05:06'),
+            ('%w',
+             '6/04/03', '%w/%U/%y', '2003-02-01T00:00:00'),
+            ('%W',  # %u requires a single digit.
+             '6/4/2003', '%u/%W/%Y', '2003-02-01T00:00:00'),
+            ('%V',
+             '6/4/2003', '%u/%V/%G', '2003-01-25T00:00:00'),
+        ]
+        def time_helper(reason, string, format, target):
+            reason = 'test single digit ' + reason
+            with self.subTest(reason=reason,
+                              string=string,
+                              format=format,
+                              target=target):
+                newdate = strptime(string, format)
+                self.assertEqual(newdate.isoformat(timespec='seconds'), target,
+                                 msg=reason)
+        for testcase in inputs:
```

Comment:
Using a closure for this seems unusual to me, did you have a particular reason for doing it this way? I can see for example that it does help to isolate the individual test cases by creating a new scope for each subtest, *but* I do find it a bit harder to read.

If you didn't have any particular reason, I'm inclined to say you should just move the body of `time_helper` into this for loop. If you have a strong reason or feel strongly about it, I think you can make the test more readable by changing the following things:

1. Move the `subtest` call out of the function, so that the function body isn't as deeply nested.

2. Rename `time_helper` to something like `run_strptime_test_case` or `assert_strptime`.

Then the loop would be something like this:

```python
for testcase in inputs:
    with self.subTest(reason=testcase[0]):
        run_strptime_test_case(*testcase)
```